src/mon/OSDMonitor.cc

```diff
@@ -8417,6 +8458,204 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
+            continue;
+          }
+
+          if (osdmap.is_noout(osd)) {
+            ss << "osd." << osd << " is already marked noout. ";
+            continue;
+          }
+
+          pending_inc.new_state[osd] |= CEPH_OSD_NOOUT;
+          any = true;
+        } else {
+          // nodown
+          if (osdmap.is_down(osd)) {
+            ss << "osd." << osd << " is already down. ";
+            continue;
+          }
+
+          if (osdmap.is_nodown(osd)) {
+            ss << "osd." << osd << " is already marked nodown. ";
+            continue;
+          }
+
+          pending_inc.new_state[osd] |= CEPH_OSD_NODOWN;
+          any = true;
+        }
+      }
+    }
+
+    if (any) {
+      getline(ss, rs);
+      wait_for_finished_proposal(op, new Monitor::C_Command(mon, op, err, rs,
+                                 get_last_committed() + 1));
+      return true;
+    }
+  } else if (prefix == "osd rm-noout" ||
+             prefix == "osd rm-nodown") {
+
+    bool any = false;
+    bool stop = false;
+    bool noout = prefix == "osd rm-noout";
+
+    vector<string> idvec;
+    cmd_getval(g_ceph_context, cmdmap, "ids", idvec);
+
+    for (unsigned j = 0; j < idvec.size() && !stop; j++) {
+      vector<int> osds;
+
+      // for removal, wildcald "any" is also supported,
+      // e.g., to quickly cancel all previous noout[nodown] osds
+      if (j == 0 && idvec[0] == "any") {
+        if (noout) {
+          osdmap.get_noout_osds(&osds);
+        } else {
+          osdmap.get_nodown_osds(&osds);
+        }
+
+        // cancel pending noout[nodown] requests too,
+        // if there is any
+        for (auto &p : pending_inc.new_state) {
+          if (noout) {
+            if (!osdmap.is_noout(p.first) &&
+                (p.second & CEPH_OSD_NOOUT) != 0) {
+              p.second &= ~CEPH_OSD_NOOUT;
+              any = true;
+            }
+          } else {
+            if (!osdmap.is_nodown(p.first) &&
+                (p.second & CEPH_OSD_NODOWN) != 0) {
+              p.second &= ~CEPH_OSD_NODOWN;
+              any = true;
+            }
+          }
+        }
+
+        stop = true;
+      } else {
+        long osd_first = -1;
+        long osd_last = -1;
+        // check if it seems like a range, e.g., "0~100"
+        auto delimeter_pos = idvec[j].find('~');
+        if (delimeter_pos != std::string::npos && // "~" exists
+          delimeter_pos != 0 && // and is neither the first char
+          delimeter_pos != idvec[j].length() - 1) { // nor the last char
+          auto first = idvec[j].substr(0, delimeter_pos);
+          auto last = idvec[j].substr(delimeter_pos + 1,
+                                      idvec[j].length() - delimeter_pos - 1);
+          osd_first = parse_osd_id(first.c_str(), &ss);
+          osd_last = parse_osd_id(last.c_str(), &ss);
+        } else {
+          // try single osd way
+          osd_first = parse_osd_id(idvec[j].c_str(), &ss);
+          osd_last = osd_first;
+        }
```

Comment:
I'm still not wild about the range syntax here.  The only time i can see thsi being useful is when you're marking the entire cluster down  or noout or whatever.  In practice, osd ids won't be sequential ranges on hosts, so I'm not sure it really helps anyone.